### PR TITLE
HID: multitouch: Support Asus T304UA media keys

### PR DESCRIPTION
**Pull request by jprvita**
The Asus T304UA convertible sports a magnetic detachable keyboard with
touchpad, which is connected over USB. Most of the keyboard hotkeys are
exposed through the same USB interface as the touchpad, defined in the
report descriptor as follows:

0x06, 0x31, 0xFF,  // Usage Page (Vendor Defined 0xFF31)
0x09, 0x76,        // Usage (0x76)
0xA1, 0x01,        // Collection (Application)
0x05, 0xFF,        //   Usage Page (Reserved 0xFF)
0x85, 0x5A,        //   Report ID (90)
0x19, 0x00,        //   Usage Minimum (0x00)
0x2A, 0xFF, 0x00,  //   Usage Maximum (0xFF)
0x15, 0x00,        //   Logical Minimum (0)
0x26, 0xFF, 0x00,  //   Logical Maximum (255)
0x75, 0x08,        //   Report Size (8)
0x95, 0x0F,        //   Report Count (15)
0xB1, 0x02,        //   Feature (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position,Non-volatile)
0x05, 0xFF,        //   Usage Page (Reserved 0xFF)
0x85, 0x5A,        //   Report ID (90)
0x19, 0x00,        //   Usage Minimum (0x00)
0x2A, 0xFF, 0x00,  //   Usage Maximum (0xFF)
0x15, 0x00,        //   Logical Minimum (0)
0x26, 0xFF, 0x00,  //   Logical Maximum (255)
0x75, 0x08,        //   Report Size (8)
0x95, 0x02,        //   Report Count (2)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC0,              // End Collection

The UsagePage is declared as a variable, but we need to treat it as an
array to be able to map each Usage we care about to its corresponding
input key.

Signed-off-by: João Paulo Rechi Vita <jprvita@endlessm.com>

https://phabricator.endlessm.com/T14926

### CODE REVIEW

File: drivers/hid/hid-multitouch.c [hunk 1]
```diff
@@ -841,6 +847,8 @@ static int mt_touch_input_configured(struct hid_device *hdev,
 	return 0;
 }
 
+#define mt_map_key_clear(c)	hid_map_usage_clear(hi, usage, bit, \
+						    max, EV_KEY, (c))
 static int mt_input_mapping(struct hid_device *hdev, struct hid_input *hi,
```
Comment (carlocaione) [inline drivers/hid/hid-multitouch.c]: nit: blank line here

File: drivers/hid/hid-multitouch.c [hunk 2]
```diff
@@ -859,10 +867,35 @@ static int mt_input_mapping(struct hid_device *hdev, struct hid_input *hi,
...
 	/*
+	 * Some Asus keyboard+touchpad devices have the hotkeys defined in the
+	 * touchpad report descriptor. We need to treat these as an array to
+	 * map usages to input keys.
+	 */
+	if (field->application == 0xff310076 &&
+	    td->mtclass.quirks & MT_QUIRK_ASUS_CUSTOM_UP &&
+	    (usage->hid & HID_USAGE_PAGE) == HID_UP_CUSTOM) {
+		set_bit(EV_REP, hi->input->evbit);
+		if (field->flags & HID_MAIN_ITEM_VARIABLE)
```
Comment (carlocaione) [inline drivers/hid/hid-multitouch.c]: can we avoid this if at all?

File: drivers/hid/hid-multitouch.c [hunk 3]
```diff
@@ -859,10 +867,35 @@ static int mt_input_mapping(struct hid_device *hdev, struct hid_input *hi,
...
+			field->flags &= ~HID_MAIN_ITEM_VARIABLE;
```
Comment (carlocaione) [inline drivers/hid/hid-multitouch.c]: nit: black line here

**Comment by carlocaione** [comment]
@jprvita I left a couple of nits and one comment.

File: drivers/hid/hid-multitouch.c (hunk 2 above)
Comment (jprvita) [inline drivers/hid/hid-multitouch.c]: yes, we could. but I switched to something else and in the meantime this got merged upstream, with a minor change (a define for 0xff310076). sorry :/

**Comment by jprvita** [comment]
@carlocaione Thanks for the review. I'll update the PR with the version which was accepted upstream.

**Pull request merged**